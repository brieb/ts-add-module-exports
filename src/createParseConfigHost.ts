import ts from "typescript";

export function createParseConfigHost(): ts.ParseConfigHost {
  return {
    useCaseSensitiveFileNames: ts.sys.useCaseSensitiveFileNames,
    fileExists: ts.sys.fileExists,
    readFile: ts.sys.readFile,
    readDirectory: ts.sys.readDirectory
  };
}
