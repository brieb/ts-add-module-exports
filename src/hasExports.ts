import astTypes from "ast-types";

const n = astTypes.namedTypes;

export interface HasExportsResult {
  hasModuleExports: boolean;
  hasExportsNamed: boolean;
  hasExportsDefault: boolean;
}

function isRootLevel(ast: any, node: any) {
  return ast.program.body.indexOf(node) !== -1;
}

function isExports(objectName: string, propertyName: string) {
  return (
    (objectName === "exports" || objectName === "_exports") &&
    propertyName !== "__esModule"
  );
}

export function hasExports(ast: any): HasExportsResult {
  let hasModuleExports = false;
  let hasExportsNamed = false;
  let hasExportsDefault = false;

  astTypes.visit(ast, {
    visitMemberExpression(path: any) {
      if (
        isRootLevel(ast, path.parent.parent.value) &&
        n.Identifier.check(path.node.object) &&
        n.Identifier.check(path.node.property)
      ) {
        const objectName = path.node.object.name;
        const propertyName = path.node.property.name;

        if (isExports(objectName, propertyName)) {
          if (propertyName === "default") {
            hasExportsDefault = true;
          } else {
            hasExportsNamed = true;
          }
        }

        if (`${objectName}.${propertyName}` === "module.exports") {
          hasModuleExports = true;
        }
      }

      this.traverse(path);
    },
    visitCallExpression(path: any) {
      if (
        isRootLevel(ast, path.parent.value) &&
        n.MemberExpression.check(path.node.callee) &&
        n.Identifier.check(path.node.callee.object) &&
        n.Identifier.check(path.node.callee.property) &&
        path.node.callee.object.name === "Object" &&
        path.node.callee.property.name === "defineProperty"
      ) {
        const obj = path.node.arguments[0];
        const prop = path.node.arguments[1];

        if (n.Identifier.check(obj) && n.Identifier.check(prop)) {
          const objectName = obj.name;
          const propertyName = prop.name;
          if (isExports(objectName, propertyName)) {
            if (propertyName === "default") {
              hasExportsDefault = true;
            } else {
              hasExportsNamed = true;
            }
          }
        }
      }

      this.traverse(path);
    }
  });

  return {
    hasModuleExports,
    hasExportsNamed,
    hasExportsDefault
  };
}
