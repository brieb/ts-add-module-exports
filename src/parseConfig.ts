import ts from "typescript";
import { createParseConfigHost } from "./createParseConfigHost";

export interface ParsedConfig {
  options: ts.CompilerOptions;
  fileNames: string[];
}

export function parseConfig(basePath: string): ParsedConfig {
  const configFileName = ts.findConfigFile(
    basePath,
    ts.sys.fileExists,
    "tsconfig.json"
  );
  if (!configFileName) {
    throw new Error("tsconfig.json file not found");
  }

  const configFileContents = ts.sys.readFile(configFileName) || "";
  const { config, error } = ts.parseConfigFileTextToJson(
    configFileName,
    configFileContents
  );
  if (error) {
    throw new Error(
      `Error parsing TS config file ${configFileName}: ${error.messageText}`
    );
  }

  const { options, fileNames, errors } = ts.parseJsonConfigFileContent(
    config,
    createParseConfigHost(),
    basePath
  );
  if (errors.length > 0) {
    throw new Error(`Errors parsing TS config: ${JSON.stringify(errors)}`);
  }

  return { options, fileNames };
}
