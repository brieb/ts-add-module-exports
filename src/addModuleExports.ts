import astTypes from "ast-types";
import { parse, print } from "recast";
import { hasExports } from "./hasExports";

const b = astTypes.builders;

export function addModuleExports(input: string): string {
  const ast = parse(input, { parser: require("recast/parsers/babel") });

  const { hasModuleExports, hasExportsNamed, hasExportsDefault } = hasExports(
    ast
  );

  if (!hasExportsNamed && !hasModuleExports && hasExportsDefault) {
    ast.program.body.push(
      b.expressionStatement(
        b.assignmentExpression(
          "=",
          b.memberExpression(b.identifier("module"), b.identifier("exports")),
          b.memberExpression(
            b.identifier("exports"),
            b.stringLiteral("default"),
            true
          )
        )
      )
    );
  }

  return print(ast).code;
}
