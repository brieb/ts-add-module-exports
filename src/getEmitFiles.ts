import ts from "typescript";
import { ParsedConfig } from "./parseConfig";

export function getEmitFiles({ options, fileNames }: ParsedConfig) {
  const program = ts.createProgram({ options, rootNames: fileNames });

  const emitFiles: string[] = [];
  program.emit(undefined, fileName => {
    emitFiles.push(fileName);
  });

  return emitFiles;
}
