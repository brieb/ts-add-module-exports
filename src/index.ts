import ts from "typescript";
import { addModuleExports } from "./addModuleExports";
import { getEmitFiles } from "./getEmitFiles";
import { parseConfig } from "./parseConfig";

const basePath = process.cwd();

const { options, fileNames } = parseConfig(basePath);

if (options.module !== ts.ModuleKind.CommonJS) {
  throw new Error(
    'Only intended for CommonJS modules (i.e. "module": "commonjs")'
  );
}

getEmitFiles({ options, fileNames })
  .filter(file => file.endsWith(".js") && ts.sys.fileExists(file))
  .forEach(jsFile => {
    ts.sys.writeFile(jsFile, addModuleExports(ts.sys.readFile(jsFile)!));
  });
