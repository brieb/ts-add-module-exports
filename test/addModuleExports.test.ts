import path from "path";
import ts from "typescript";
import { parseConfig } from "../src/parseConfig";
import { addModuleExports } from "../src/addModuleExports";

describe("addModuleExports", () => {
  it("Properly transforms demo", () => {
    const basePath = path.resolve(__dirname, "../demo");
    const { options, fileNames } = parseConfig(basePath);
    const program = ts.createProgram({ options, rootNames: fileNames });

    const inputs: { [fileName: string]: string } = {};
    program.emit(undefined, (fileName, data) => {
      const relFileName = path.relative(basePath, fileName);
      inputs[relFileName] = data;
    });

    const outputFileNames = Object.keys(inputs)
      .filter(fileName => fileName.endsWith(".js"))
      .sort();
    expect(outputFileNames).toMatchSnapshot();

    outputFileNames.forEach(fileName => {
      expect(addModuleExports(inputs[fileName])).toMatchSnapshot(fileName);
    });
  });
});
