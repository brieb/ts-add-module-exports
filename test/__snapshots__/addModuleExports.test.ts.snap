// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`addModuleExports Properly transforms demo 1`] = `
Array [
  "out/root-level-file.js",
  "out/src/export-default.js",
  "out/src/export-equals-obj.js",
  "out/src/export-equals.js",
  "out/src/export-named.js",
]
`;

exports[`addModuleExports Properly transforms demo: out/root-level-file.js 1`] = `
"\\"use strict\\";;
Object.defineProperty(exports, \\"__esModule\\", { value: true });
function fn() { }
exports.default = fn;
module.exports = exports[\\"default\\"];
"
`;

exports[`addModuleExports Properly transforms demo: out/src/export-default.js 1`] = `
"\\"use strict\\";;
Object.defineProperty(exports, \\"__esModule\\", { value: true });
function main() {
    console.log(\\"main\\");
}
exports.default = main;
module.exports = exports[\\"default\\"];
"
`;

exports[`addModuleExports Properly transforms demo: out/src/export-equals.js 1`] = `
"\\"use strict\\";
function main() {
    console.log(\\"main\\");
}
module.exports = main;
"
`;

exports[`addModuleExports Properly transforms demo: out/src/export-equals-obj.js 1`] = `
"\\"use strict\\";
function main() {
    console.log(\\"main\\");
}
module.exports = {
    main: main
};
"
`;

exports[`addModuleExports Properly transforms demo: out/src/export-named.js 1`] = `
"\\"use strict\\";
Object.defineProperty(exports, \\"__esModule\\", { value: true });
function main() {
    console.log(\\"main\\");
}
exports.default = main;
function named() {
    console.log(\\"named\\");
}
exports.named = named;
"
`;
